### PR TITLE
feat: add documentation details to endpoints

### PR DESCRIPTION
**Pull request by ikadix**
* endpoint `path` has any `:` removed from url params.
As we are not using these and the doc generator seems to break on them.
* endpoint `name` becomes `summary`
* endpoint `description` becomes `description`
	* if endpoint has scopes they are added to the description
		An example of the resulting description.
		```MD
		Returns a list of all global address lists.

		## Scopes

		- `address_lists`
		-  `address_lists:read`
		```
		
* endpoint `group.name` becomes the `tag` if present, else name (`Core`) is still used.
* At the top level `tags` and `x-tagGroups` have been added.
	*  Tags are the route groups name and a hierarchy is built based on group parents.
	* If tag is not in any route groups after they've been parsed, a tag group is created just containing that tag.

### CODE REVIEW

File: examples/core_api/controllers/time_controller.rb [hunk 1]
```diff
@@ -19,7 +19,9 @@ class TimeController < Apia::Controller
       # TODO: add example of multiple objects using the same objects, to ensure
       # we are handing circular references correctly
       endpoint :format do
+        name "Format Time"
         description "Format the given time"
+        scopes "time", "time:format"
```
Comment (paulsturgess) [inline examples/core_api/controllers/time_controller.rb:24]: Thanks for updating these to ensure our tests cover the new functionality 👍

File: lib/apia/open_api/objects/path.rb [hunk 2]
```diff
@@ -29,27 +29,64 @@ class Path
...
           @api_authenticator = api_authenticator
           @route_spec = {
             operationId: convert_route_to_id,
-            tags: [name]
+            summary: @route.endpoint.definition.name,
+            description: @route.endpoint.definition.description,
+            tags: route.group ? get_group_name(route.group) : [name],
+            security: [{
+              Authenticator: @route.endpoint.definition.scopes
+
```
[the remark this answers is not in the document]
Comment (paulsturgess) [reply lib/apia/open_api/objects/path.rb:56]: I also wonder if we should omit `security` if the endpoint doesn't have any scopes? But maybe it doesn't do any harm 🤷

File: lib/apia/open_api/objects/path.rb [hunk 3]
```diff
@@ -29,27 +29,64 @@ class Path
...
           else
             add_request_body
           end
 
-          @spec[:paths]["/#{path}"] ||= {}
-          @spec[:paths]["/#{path}"][@route.request_method.to_s] = @route_spec
+          path = "/#{path}"
+          # Remove the `:` from the url parameters in the path
+          # This is because some tools based on the OpenAPI spec don't like the `:` in the path
+          path = path.gsub(/:([^\/]+)/, '\1')
```
Comment (paulsturgess) [inline lib/apia/open_api/objects/path.rb:86]: 👍

File: lib/apia/open_api/specification.rb [hunk 4]
```diff
@@ -85,6 +96,55 @@ def add_security
...
+
+        @spec[:tags].each do |tag|
+          unless @spec[:"x-tagGroups"].any? { |group| group[:tags].include?(tag[:name]) }
+            @spec[:"x-tagGroups"] << { name: tag[:name], tags: [tag[:name]] }
+          end
+        end
+
+        @spec[:"x-tagGroups"].sort_by! { |group| group[:name] }
+        @spec[:"x-tagGroups"].each { |group| group[:tags].sort! }
+      end
```
Comment (paulsturgess) [inline lib/apia/open_api/specification.rb:146]: I won't pretend to have verified the output here. But I'm assuming you know the format you want.

Comment (ikadix) [reply lib/apia/open_api/specification.rb:146]: I have tested this gives the desired output manually but I'll be honest this is the worst part of this PR. 
Anything that stands out as odd or bad in particular?

File: lib/apia/open_api/objects/path.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (ikadix) [reply lib/apia/open_api/objects/path.rb:56]: Oh I don't think we need this after adding the scopes to the description.
We should probably hold back to add scopes in security on endpoints as part of introducing the oauth2 security scheme.

**Review by paulsturgess — APPROVED** [review]
Nice one @ikadix 

For some reason the specs fail locally for me. Just cause the paths are in a different order, the content it the same. 🤷 

Maybe we should order the paths alphabetically:

```ruby
def sorted_json
  @spec[:paths] = @spec[:paths].sort_by { |path, _| path }.to_h
  JSON.pretty_generate(@spec)
end
```

But that could be introduced in another PR.

Or we could just order alphabetically for the purposes of the spec. But that obviously introduces a risk that we're not checking the output of the actual spec we will publish. 😬

**Pull request merged**